smarts/core/chassis.py

```diff
@@ -440,26 +441,46 @@ def __init__(
         if initial_speed is not None:
             self._initialize_speed(initial_speed)
 
-    @property
+    @cached_property
+    def _cached_props(self):
+        return {
+            a
+            for a, v in self.__class__.__dict__.items()
+            if isinstance(v, cached_property)
+        }
+
+    def _clear_step_cache(self):
+        cached_props = self._cached_props
+        self.__dict__ = {
+            a: v
+            for a, v in self.__dict__.items()
+            if a not in cached_props or a == "_cached_props"
+        }
+
+    @cached_property
     def pose(self) -> Pose:
         pos, orn = self._client.getBasePositionAndOrientation(self._bullet_id)
         heading = Heading(yaw_from_quaternion(orn))
-
-        pose = Pose.from_explicit_offset(
-            [0, 0, 0],
-            np.array(pos),
-            heading,
-            local_heading=Heading(0),
-        )
-        return pose
+        if not self._pose:
+            self._pose = Pose.from_explicit_offset(
+                [0, 0, 0],
+                np.array(pos),
+                heading,
+                local_heading=Heading(0),
+            )
+        else:
+            self._pose.reset_with(pos, heading)
```

Comment:
Of all of the changes in this PR, this one was probably the most risky.
Instead of creating a new `Pose` object on every (sub)step of pybullet (indeed, on every single call to the `pose` property!), we now reuse the same object over and over.  

This is a huge win in terms of efficiency and performance, but it carries that risk that anywhere that a `Pose` object reference was stored and kept for later use might now be dealing with a changed object!

I did a bit of an audit of this, and I think things are ok (all of the tests now pass), but it was the reason for the ~2 test failures that I encountered along the way.  This is why it was important to actually _copy_ the `Pose` and/or its `position` field before returning it in the `Observations` class.

The change made to also do this before emitting states to envision was technically un-necessary, but was made necessary by the way the data replay test was written (specifically the way the websocket was mocked), but I made it anyway, just to be consistent.

Reply:
The external mutability of the pose object is a bit concerning but I do agree that it is a good win for efficiency.